aw_watcher_window/macos.swift

```diff
@@ -0,0 +1,285 @@
+import Cocoa
+import ScriptingBridge
+
+@objc protocol ChromeTab {
+  @objc optional var URL: String { get }
+  @objc optional var title: String { get }
+}
+
+@objc protocol ChromeWindow {
+  @objc optional var activeTab: ChromeTab { get }
+  @objc optional var mode: String { get }
+}
+
+extension SBObject: ChromeWindow, ChromeTab {}
+
+@objc protocol ChromeProtocol {
+  @objc optional func windows() -> [ChromeWindow]
+}
+
+extension SBApplication: ChromeProtocol {}
+
+struct NetworkMessage: Codable {
+  var app: String
+  var title: String
+  var url: String?
+}
+
+struct Heartbeat: Codable {
+  var timestamp: Date
+  var data: NetworkMessage
+}
+
+struct Bucket: Codable {
+  var client: String
+  var type: String
+  var hostname: String
+}
+
+let clientHostname = ProcessInfo.processInfo.hostName
+let clientName = "aw-watcher-window2"
+let bucketName = "\(clientName)_\(clientHostname)"
+
+let main = MainThing()
+var oldHeartbeatData: NetworkMessage?
+
+let encoder = JSONEncoder()
+let formatter = ISO8601DateFormatter()
+formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
+
+encoder.dateEncodingStrategy = .custom({ date, encoder in
+  var container = encoder.singleValueContainer()
+  let dateString = formatter.string(from: date)
+  try container.encode(dateString)
+})
+
+start()
+RunLoop.main.run()
+
+func start() {
+  guard checkAccess() else {
+    DispatchQueue.main.asyncAfter(deadline: .now() + 10) {
+      start()
+    }
+    return
+  }
+
+  createBucket()
+
+  NSWorkspace.shared.notificationCenter.addObserver(
+    main, selector: #selector(main.focusedAppChanged),
+    name: NSWorkspace.didActivateApplicationNotification,
+    object: nil)
+  main.focusedAppChanged()
+  detectIdle()
+}
+
+func detectIdle() {
+  let seconds = 15.0 - SystemIdleTime()!
+  if seconds < 0.0 {
+    sendHeartbeat(data: NetworkMessage(app: "", title: ""))
```

Comment:
Why send an empty one? Won't this cause issues?

Reply:
The way I've structured this is to send an empty heartbeat as a representation of being idle because that gives precise timing of when your computer is idle instead of just when the pulseTime expires.

The swift code is designed to always send a message on change, so what happens is when you change apps it immediately sends a heartbeat to extend the last network request until now, then it sends a heartbeat to start the updated request. Because of that, I need to send something to signify that you are going idle immediately, and an empty title displays as shorter in the ui which works nicely (maybe I could send pulseTime=0 for idle instead?)

Reply:
> maybe I could send pulseTime=0 for idle instead?

Yes, if I understand you correctly I think this would accomplish the same thing.

Also, if you detect the idle state after 15s, you should send a idle-heartbeat starting 15s ago (or whenever activity was last detected), not after the timeout has passed.